File: Sum/8.string_to_integer.dart
class Solution{
  int myAtoi(String s){
    s = s.trim();
    if (s.isEmpty) return 0;
    
    int i = 0;
    int sign = 1;
    int result = 0;
    
    if(s[i] == '-' || s[i] == '+'){
      sign = (s[i] == '-') ? -1 : 1;
      i++;
    }
    
    // codeUnitAt এর কাজ হচ্ছে Unicode এর মান বের করা 
    while (i < s.length && s[i].codeUnitAt(0) >= '0'.codeUnitAt(0) && s[i].codeUnitAt(0) <= '9'.codeUnitAt(0)) { // if (c.codeUnitAt(0) >= 48 && c.codeUnitAt(0) <= 57)
      
      // '7' → 55 (ASCII) -> s[i].codeUnitAt(0)
      // '0' → 48 ->  '0'.codeUnitAt(0)
      
      int digit = s[i].codeUnitAt(0) - '0'.codeUnitAt(0); // digit = 55 - 48 = 7 → ✅ digit 7
      
      // Check for overflow (32-bit signed int)
      if (result > (2147483647 - digit) ~/ 10) {
        return (sign == -1) ? -2147483648 : 2147483647;
      }
      
      result = result * 10 + digit;
      i++;
    }
    
    return result * sign;
  }
}

// void main(){
//   final s = Solution();
//   final result = s.myAtoi(" -1337c0d3");
  
//   print(result);
// }

File: Sum/Python/8. string_to_integer.py
class Solution:
    def myAtoi(self, s: str)-> int:
        s = s.lstrip()
        if not s: return 0
        
        sign = 1
        i = 0
        
        if s[i] in ['+', '-']:
            sign = -1 if s[i] == '-' else 1
            i += 1
            
        result = 0
        while i < len(s) and s[i].isdigit():
            digit = int(s[i])
            
            if result > (2**31 - 1 - digit) // 10:
                return -2**31 if sign == -1 else 2**31 - 1
                
            result = result * 10 + digit
            
            # result = 0
            # digit = 1 → result = 1
            # digit = 3 → result = 13
            # digit = 3 → result = 133
            # digit = 7 → result = 1337
            
            i += 1
            
        return result * sign

# s = Solution()
# result = s.myAtoi('2147483648')
# print(result)